File: benchmarks/baselines/go/v2/store/store.go
// Package store provides the deterministic jobs capability used by the benchmark.
package store

import (
	"bytes"
	"fmt"

	"ail.dev/job-service-baseline/v2/domain"
)

type RecordVersion uint8

const (
	RecordVersionV1 RecordVersion = 1
	RecordVersionV2 RecordVersion = 2
)

// StoredJob preserves the declared persisted schema version.
type StoredJob struct {
	RecordVersion RecordVersion
	JobID         string
	Task          string
	Payload       []byte
	Priority      *domain.Priority
}

// AdaptToV2 explicitly maps a V1 persisted record to normal priority.
func (stored StoredJob) AdaptToV2() domain.Job {
	priority := domain.PriorityNormal
	if stored.RecordVersion == RecordVersionV2 && stored.Priority != nil {
		priority = *stored.Priority
	}
	return domain.Job{
		JobID:    stored.JobID,
		Task:     stored.Task,
		Payload:  bytes.Clone(stored.Payload),
		Priority: priority,
	}
}

type StoreCall struct {
	Job StoredJob
}

// DeterministicJobStore records every call and applies a supplied outcome.
type DeterministicJobStore struct {
	jobs          []StoredJob
	outcome       domain.InsertOutcome
	insertVersion RecordVersion
	calls         []StoreCall
}

func New(
	jobs []StoredJob,
	outcome domain.InsertOutcome,
	insertVersion RecordVersion,
) *DeterministicJobStore {
	return &DeterministicJobStore{
		jobs:          cloneStoredJobs(jobs),
		outcome:       outcome,
		insertVersion: insertVersion,
	}
}

func (store *DeterministicJobStore) Jobs() []StoredJob {
	return cloneStoredJobs(store.jobs)
}

func (store *DeterministicJobStore) Calls() []StoreCall {
	result := make([]StoreCall, len(store.calls))
	for index, call := range store.calls {
		result[index] = StoreCall{Job: cloneStoredJob(call.Job)}
	}
	return result
}

func (store *DeterministicJobStore) InsertIfAbsent(job domain.Job) domain.InsertOutcome {
	stored := store.encodeForInsert(job)
	store.calls = append(store.calls, StoreCall{Job: cloneStoredJob(stored)})
	if store.outcome == domain.Inserted {
		for _, current := range store.jobs {
			if current.JobID == job.JobID {
				panic(fmt.Sprintf(
					"inserted outcome violates insert-if-absent postcondition for %q",
					job.JobID,
				))
			}
		}
		store.jobs = append(store.jobs, stored)
	}
	return store.outcome
}

func (store *DeterministicJobStore) encodeForInsert(job domain.Job) StoredJob {
	stored := StoredJob{
		RecordVersion: store.insertVersion,
		JobID:         job.JobID,
		Task:          job.Task,
		Payload:       bytes.Clone(job.Payload),
	}
	if store.insertVersion == RecordVersionV2 {
		priority := job.Priority
		stored.Priority = &priority
	}
	return stored
}

func cloneStoredJobs(jobs []StoredJob) []StoredJob {
	result := make([]StoredJob, len(jobs))
	for index, job := range jobs {
		result[index] = cloneStoredJob(job)
	}
	return result
}

func cloneStoredJob(job StoredJob) StoredJob {
	job.Payload = bytes.Clone(job.Payload)
	if job.Priority != nil {
		priority := *job.Priority
		job.Priority = &priority
	}
	return job
}

File: benchmarks/baselines/rust/v2/src/store.rs
use crate::domain::{InsertOutcome, Job, JobStore, Priority};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RecordVersion {
    V1,
    V2,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum StoredJob {
    V1 {
        job_id: String,
        task: String,
        payload: Vec<u8>,
    },
    V2 {
        job_id: String,
        task: String,
        payload: Vec<u8>,
        priority: Priority,
    },
}

impl StoredJob {
    #[must_use]
    pub fn job_id(&self) -> &str {
        match self {
            Self::V1 { job_id, .. } | Self::V2 { job_id, .. } => job_id,
        }
    }

    #[must_use]
    pub fn adapt_to_v2(&self) -> Job {
        match self {
            Self::V1 {
                job_id,
                task,
                payload,
            } => Job {
                job_id: job_id.clone(),
                task: task.clone(),
                payload: payload.clone(),
                priority: Priority::Normal,
            },
            Self::V2 {
                job_id,
                task,
                payload,
                priority,
            } => Job {
                job_id: job_id.clone(),
                task: task.clone(),
                payload: payload.clone(),
                priority: *priority,
            },
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct StoreCall {
    pub job: StoredJob,
}

#[derive(Debug)]
pub struct DeterministicJobStore {
    jobs: Vec<StoredJob>,
    outcome: InsertOutcome,
    insert_version: RecordVersion,
    calls: Vec<StoreCall>,
}

impl DeterministicJobStore {
    #[must_use]
    pub fn new(
        jobs: Vec<StoredJob>,
        outcome: InsertOutcome,
        insert_version: RecordVersion,
    ) -> Self {
        Self {
            jobs,
            outcome,
            insert_version,
            calls: Vec::new(),
        }
    }

    #[must_use]
    pub fn jobs(&self) -> &[StoredJob] {
        &self.jobs
    }

    #[must_use]
    pub fn calls(&self) -> &[StoreCall] {
        &self.calls
    }

    fn encode_for_insert(&self, job: &Job) -> StoredJob {
        match self.insert_version {
            RecordVersion::V1 => StoredJob::V1 {
                job_id: job.job_id.clone(),
                task: job.task.clone(),
                payload: job.payload.clone(),
            },
            RecordVersion::V2 => StoredJob::V2 {
                job_id: job.job_id.clone(),
                task: job.task.clone(),
                payload: job.payload.clone(),
                priority: job.priority,
            },
        }
    }
}

impl JobStore for DeterministicJobStore {
    fn insert_if_absent(&mut self, job: &Job) -> InsertOutcome {
        let stored = self.encode_for_insert(job);
        self.calls.push(StoreCall {
            job: stored.clone(),
        });
        if self.outcome == InsertOutcome::Inserted {
            assert!(
                self.jobs
                    .iter()
                    .all(|current| current.job_id() != job.job_id),
                "inserted outcome violates insert-if-absent postcondition"
            );
            self.jobs.push(stored);
        }
        self.outcome
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn job(priority: Priority) -> Job {
        Job {
            job_id: "job-1".into(),
            task: "task".into(),
            payload: vec![1, 2, 3],
            priority,
        }
    }

    #[test]
    fn v1_records_adapt_explicitly_to_normal() {
        let record = StoredJob::V1 {
            job_id: "legacy".into(),
            task: "task".into(),
            payload: vec![1],
        };
        assert_eq!(record.adapt_to_v2().priority, Priority::Normal);
    }

    #[test]
    fn inserted_job_is_recorded_and_persisted_exactly_once() {
        let mut store =
            DeterministicJobStore::new(Vec::new(), InsertOutcome::Inserted, RecordVersion::V2);
        let job = job(Priority::High);

        assert_eq!(store.insert_if_absent(&job), InsertOutcome::Inserted);
        assert_eq!(store.calls().len(), 1);
        assert_eq!(store.jobs().len(), 1);
        assert_eq!(store.jobs()[0].adapt_to_v2(), job);
    }

    #[test]
    fn duplicate_and_unavailable_outcomes_preserve_initial_state() {
        let initial = StoredJob::V2 {
            job_id: "existing".into(),
            task: "old".into(),
            payload: Vec::new(),
            priority: Priority::Low,
        };
        for outcome in [
            InsertOutcome::Duplicate,
            InsertOutcome::UnavailableBeforeCommit,
        ] {
            let mut store =
                DeterministicJobStore::new(vec![initial.clone()], outcome, RecordVersion::V2);
            assert_eq!(store.insert_if_absent(&job(Priority::High)), outcome);
            assert_eq!(store.jobs(), &[initial.clone()]);
            assert_eq!(store.calls().len(), 1);
        }
    }
}
